### PR TITLE
feat(shim-sev): Implement get_attestation for SEV

### PR DESCRIPTION
**Pull request by haraldh**
Inject the secret and provide it with the get_attestation call.

Fixes: https://github.com/enarx/enarx-keepldr/issues/159

### CODE REVIEW

File: internal/shim-sev/src/syscall.rs [hunk 1]
```diff
@@ -205,11 +208,29 @@ impl SyscallHandler for Handler {
...
+
+        match SEV_SECRET.read().deref() {
+            Some(secret) => {
+                let mut result_len = size_of::<SevSecret>();
+
+                if buf_len != 0 {
+                    result_len = result_len.min(buf_len);
+                    let buf = buf.validate_slice(buf_len, self).ok_or(libc::EFAULT)?;
+
+                    buf[..result_len].copy_from_slice(&secret.data[..result_len]);
```
Comment (connorkuehl) [inline internal/shim-sev/src/syscall.rs:224]: Is the expectation here that userspace knows exactly how long the secret is?

File: src/backend/kvm/vm/builder.rs [hunk 2]
```diff
@@ -38,7 +38,17 @@ pub trait Hook {
...
+    fn code_loaded(
+        &mut self,
+        _vm: &mut VmFd,
+        _addr_space: &[u8],
+        secret: *mut SevSecret,
+    ) -> Result<()> {
+        // Inject dummy secret
+        unsafe {
+            secret.write_bytes(0xEA, 1);
+        }
```
Comment (connorkuehl) [inline src/backend/kvm/vm/builder.rs:50]: KVM should return ENOSYS for `get_attestation()`, so I'm not sure the hypervisor should plant anything here in that case.

File: tests/bin/kvm_get_att_quote.c [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    if ((size < 0) || (technology != 0))
+        return -1;
+
+    for (i = 0; i < sizeof(buf); i++) {
+        if (buf[i] != 0xEA)
+	    return 1;
+    }
+
+    return 0;
+}
```
Comment (connorkuehl) [inline tests/bin/kvm_get_att_quote.c:26]: Same comment above, `get_attestation()` shouldn't be implemented for KVM.

File: tests/bin/sev_get_att_quote.c [hunk 4]
```diff
@@ -0,0 +1,34 @@
+#include "libc.h"
+#include <errno.h>
+
+/* This test will be run only for KVM. It is designed to request a
```
Comment (connorkuehl) [inline tests/bin/sev_get_att_quote.c:4]: ~~KVM~~ SEV

File: tests/bin/sev_get_att_quote.c [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+ * Quote from get_attestation() and check that the first bytes of
+ * the returned Quote in buf match expected values. */
+
+int main(void) {
```
Comment (connorkuehl) [inline tests/bin/sev_get_att_quote.c:8]: This test will soon be "baked in" to the unattested launch feature I'm working on. We can't easily do the remote attestation protocol within our barebones libc here. I think this file can be removed.

File: tests/integration_tests.rs [hunk 6]
```diff
@@ -260,6 +260,19 @@ fn sgx_get_att_quote() {
     run_test("sgx_get_att_quote", 0, None, None, None);
 }
 
+#[cfg(feature = "backend-kvm")]
+#[cfg(not(feature = "backend-sev"))]
```
Comment (connorkuehl) [inline tests/integration_tests.rs:264]: These save us some bytes in the compiled binary but this won't prevent someone from accidentally running this with the wrong backend.

**Review by connorkuehl — CHANGES_REQUESTED** [review]

File: tests/integration_tests.rs (hunk 6 above)
Comment (haraldh) [reply tests/integration_tests.rs:264]: right

File: tests/bin/sev_get_att_quote.c (hunk 5 above)
Comment (haraldh) [reply tests/bin/sev_get_att_quote.c:8]: ok

File: tests/bin/sev_get_att_quote.c (hunk 4 above)
Comment (haraldh) [reply tests/bin/sev_get_att_quote.c:4]: removed the file

File: tests/bin/kvm_get_att_quote.c (hunk 3 above)
Comment (haraldh) [reply tests/bin/kvm_get_att_quote.c:26]: yeah

File: src/backend/kvm/vm/builder.rs (hunk 2 above)
Comment (haraldh) [reply src/backend/kvm/vm/builder.rs:50]: ok

File: internal/shim-sev/src/syscall.rs (hunk 1 above)
Comment (haraldh) [reply internal/shim-sev/src/syscall.rs:224]: for now.. yes

File: .github/workflows/test.yml [hunk 7]
```diff
@@ -27,7 +27,7 @@ jobs:
       - uses: actions-rs/cargo@v1
         with:
           command: test
-          args: ${{ matrix.profile.flag }} --features=backend-${{ matrix.backend.name }}
+          args: ${{ matrix.profile.flag }} --no-default-features --features=backend-${{ matrix.backend.name }}
```
[the remark this answers is not in the document]
Comment (haraldh) [reply .github/workflows/test.yml]: yes

**Comment by haraldh** [comment]
@connorkuehl updated

**Comment by connorkuehl** [comment]
This all looks good to me. Each time we get closer to an MVP it just further reinforces the point that I think perhaps our VM builder platform could be a little more modular, otherwise it just looks like it starts to "grow" things. I'm cogitating on that.

I'm wondering if at least in the hypervisor side the function parameters could be references instead of pointers, but I haven't looked around at the surrounding areas to see if the borrow checker would allow that.

**Comment by haraldh** [comment]
Needs https://github.com/enarx/enarx-keepldr/pull/188 first

**Comment by haraldh** [comment]
Needs https://github.com/enarx/enarx-keepldr/pull/188 and https://github.com/enarx/enarx-keepldr/pull/192 first

**Comment by haraldh** [comment]
Rebased to current master.

**Comment by haraldh** [comment]
@connorkuehl please re-review... you are still marked with `requested changes`

**Review by connorkuehl — APPROVED** [review]

**Review by lkatalin — APPROVED** [review]
I'm not an expert on some of these details (esp. CBOR length), but looks okay to me.

**Pull request merged**